cassovary-core/src/main/scala/com/twitter/cassovary/graph/ConcurrentHashMapDynamicGraph.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2015 TODO
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.twitter.cassovary.graph
+
+import com.twitter.cassovary.graph.StoredGraphDir._
+import com.twitter.cassovary.graph.node.DynamicNode
+
+/**
+ * An efficient dynamic graph implementation which supports concurrent reading and writing.  Locks are only used by writing
+ * threads, which improves efficiency.
+ * Nodes are stored in a ConcurrentHashMap, and neighbors of each node are stored in Array[Int]s.  Currently, only edge
+ * and node addition (not deletion) is supported.
+ */
+class ConcurrentHashMapDynamicGraph(storedGraphDir: StoredGraphDir = BothInOut)
+    extends DynamicDirectedGraphHashMap(storedGraphDir) {
+  override def nodeFactory(id: Int): DynamicNode = new ConcurrentNode(id)
+  override def removeEdge(srcId: Int, destId: Int) = throw new UnsupportedOperationException()
+}
+
+private class ConcurrentNode(val id: Int) extends DynamicNode {
+  val outboundList = new ConcurrentIntArrayList()
+  val inboundList = new ConcurrentIntArrayList()
+
+  override def outboundNodes(): Seq[Int] = outboundList.toSeq
```

Comment:
In high performance client code, if they want multiple random neighbors, they could still write
val outNeighbors = node.outboundNodes()
outNeighbors(5)
outNeighbors(10)
...
I would guess that most random neighbor accesses will be random walk generation, which only looks at one random neighbor from each node it walks to.

I feel somewhat nervous about lazy val because it doesn't update (as you mentioned) which could create bizarre behavior and it hides some details that could be important.